worker's error log: StatusRuntimeException: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size

server's config file
[sankuai@0df5ae3539d3 bazel-buildfarm]$ cat examples/server.config
instances {
  name: "default_memory_instance"
  digest_function: SHA256
  memory_instance_config: {
    list_operations_default_page_size: 1024
    list_operations_max_page_size: 16384
    tree_default_page_size: 1024
    tree_max_page_size: 16384
    operation_poll_timeout: {
      seconds: 30
      nanos: 0
    }
    operation_completed_delay: {
      seconds: 10
      nanos: 0
    }

    cas_config: {
      memory: {
      max_size_bytes: 1073741824 # 1024 * 1024 * 1024
      }
    }

    action_cache_config: {
      delegate_cas: {}
    }

    default_action_timeout: {
      seconds: 600
      nanos: 0
    }

   maximum_action_timeout: {
      seconds: 3600
      nanos: 0
    }
  }
}

instances {
  name: "another_memory_instance"

  digest_function: SHA1

  memory_instance_config: {
    list_operations_default_page_size: 1024
    list_operations_max_page_size: 16384

    tree_default_page_size: 1024
    tree_max_page_size: 16384

    operation_poll_timeout: {
      seconds: 30
      nanos: 0
    }

    operation_completed_delay: {
      seconds: 10
      nanos: 0
    }

    cas_config: {
      grpc: {
        instance_name: "internal"

        target: "cas-host.cloud.org:5000"
      }
    }

    action_cache_config: {
      grpc: {
        instance_name: "lru-ac"

        target: "ac-host.cloud.org:5001"
      }
    }

    default_action_timeout: {
      seconds: 600
      nanos: 0
    }

    maximum_action_timeout: {
      seconds: 3600
      nanos: 0
    }
  }
}

port: 8980

default_instance_name: "default_memory_instance"`

worker's config file
`[sankuai@0df5ae3539d3 bazel-buildfarm]$ cat examples/worker.config

digest_function: SHA256

operation_queue: {
  target: "localhost:8980"

  instance_name: "default_memory_instance"

  deadline_after_seconds: 60
}

content_addressable_storage: {
  target: "localhost:8980"

  instance_name: "default_memory_instance"

  deadline_after_seconds: 60
}

action_cache: {
  target: "localhost:8980"

  instance_name: "default_memory_instance"

  deadline_after_seconds: 60
}

root: "/tmp/worker"

cas_cache_directory: "cache"
inline_content_limit: 1048567 # 1024 * 1024

stream_stdout: true

stdout_cas_policy: ALWAYS_INSERT

stream_stderr: true

stderr_cas_policy: ALWAYS_INSERT

above
file_cas_policy: ALWAYS_INSERT

tree_page_size: 0

operation_poll_period: {
  seconds: 1
  nanos: 0
}

platform: {
}

cas_cache_max_size_bytes: 2147483648 # 2 * 1024 * 1024 * 1024

cas_cache_max_entry_size_bytes: 2147483648 # 2 * 1024 * 1024 * 1024

execute_stage_width: 1

input_fetch_stage_width: 1

default_action_timeout: {
  seconds: 600
  nanos: 0
}

maximum_action_timeout: {
  seconds: 3600
  nanos: 0
}`

logs of server
`INFO: REQUEUEING default_memory_instance/operations/a192bcc9-b238-4d8b-a48c-4a3487afb33d [Thu Jan 09 12:12:33 UTC 2020]
INFO: default_memory_instance::requeueOperation(ed7fa470dba4c07a1710ef0799b9955a4fb2717a4a17254891f15398bf4a81e2/141): default_memory_instance/operations/a192bcc9-b238-4d8b-a48c-4a3487afb33d [Thu Jan 09 12:12:33 UTC 2020]
Exception in thread "Thread-1792" java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:492)
	at build.buildfarm.instance.memory.MemoryInstance.enqueueOperation(MemoryInstance.java:379)
	at build.buildfarm.instance.AbstractServerInstance.putOperation(AbstractServerInstance.java:1298)
	at build.buildfarm.instance.memory.MemoryInstance.putOperation(MemoryInstance.java:508)
	at build.buildfarm.instance.AbstractServerInstance.requeueOperation(AbstractServerInstance.java:1458)
	at build.buildfarm.instance.memory.MemoryInstance.lambda$onDispatched$4(MemoryInstance.java:588)
	at build.buildfarm.common.Watchdog.run(Watchdog.java:47)
	at java.base/java.lang.Thread.run(Thread.java:834)`

logs of worker
`WARNING: error downloading cfb460de773900ff6ff856846303ca4a4e04add581c2f890a6c7f62b91cecc74/4489118
java.io.IOException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size 4194304: 4489123
	at build.buildfarm.common.io.ByteStringQueueInputStream.read(ByteStringQueueInputStream.java:80)
	at build.buildfarm.common.io.ByteStringQueueInputStream.read(ByteStringQueueInputStream.java:65)
	at com.google.common.io.ByteStreams.copy(ByteStreams.java:108)
	at build.buildfarm.worker.CASFileCache.copyExternalInput(CASFileCache.java:1831)
	at build.buildfarm.worker.CASFileCache.putAndCopy(CASFileCache.java:1801)
	at build.buildfarm.worker.CASFileCache.lambda$put$19(CASFileCache.java:1783)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:213)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:202)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:118)
	at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:982)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:261)
	at com.google.common.util.concurrent.MoreExecutors$5.execute(MoreExecutors.java:977)
	at com.google.common.util.concurrent.ImmediateFuture.addListener(ImmediateFuture.java:39)
	at com.google.common.util.concurrent.AbstractTransformFuture.create(AbstractTransformFuture.java:39)
	at com.google.common.util.concurrent.Futures.transformAsync(Futures.java:471)
	at build.buildfarm.worker.CASFileCache.put(CASFileCache.java:1780)
	at build.buildfarm.worker.CASFileCache.putDirectoryFiles(CASFileCache.java:1495)
	at build.buildfarm.worker.CASFileCache.putFiles(CASFileCache.java:1478)
	at build.buildfarm.worker.operationqueue.Worker.fetchInputs(Worker.java:267)
	at build.buildfarm.worker.operationqueue.Worker.fetchInputs(Worker.java:280)
	at build.buildfarm.worker.operationqueue.Worker.fetchInputs(Worker.java:280)
	at build.buildfarm.worker.operationqueue.Worker.fetchInputs(Worker.java:280)
	at build.buildfarm.worker.operationqueue.Worker.access$700(Worker.java:122)
	at build.buildfarm.worker.operationqueue.Worker$2.createExecDir(Worker.java:680)
	at build.buildfarm.worker.InputFetcher.fetchPolled(InputFetcher.java:105)
	at build.buildfarm.worker.InputFetcher.runInterruptibly(InputFetcher.java:78)
	at build.buildfarm.worker.InputFetcher.run(InputFetcher.java:169)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size 4194304: 4489123
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:449)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:521)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:641)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:529)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:703)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:692)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more`

--- a/src/main/java/build/buildfarm/server/ByteStreamService.java
+++ b/src/main/java/build/buildfarm/server/ByteStreamService.java
@@ -63,7 +63,7 @@ import java.util.logging.Logger;
 public class ByteStreamService extends ByteStreamImplBase {
   private static final Logger logger = Logger.getLogger(ByteStreamService.class.getName());
 
-  private static int CHUNK_SIZE = 64 * 1024;
+  static int CHUNK_SIZE = 64 * 1024;
 
   private final long deadlineAfter;
   private final TimeUnit deadlineAfterUnits;
@@ -211,9 +211,19 @@ public class ByteStreamService extends ByteStreamImplBase {
     return new DelegateServerCallStreamObserver<ByteString, ReadResponse>(responseObserver) {
       @Override
       public void onNext(ByteString data) {
-        responseObserver.onNext(ReadResponse.newBuilder()
-            .setData(data)
-            .build());
+        while (!data.isEmpty()) {
+          ByteString slice;
+          if (data.size() > CHUNK_SIZE) {
+            slice = data.substring(0, CHUNK_SIZE);
+            data = data.substring(CHUNK_SIZE);
+          } else {
+            slice = data;
+            data = ByteString.EMPTY;
+          }
+          responseObserver.onNext(ReadResponse.newBuilder()
+              .setData(slice)
+              .build());
+        }
       }
 
       @Override

--- a/src/test/java/build/buildfarm/server/ByteStreamServiceTest.java
+++ b/src/test/java/build/buildfarm/server/ByteStreamServiceTest.java
@@ -15,6 +15,7 @@
 package build.buildfarm.server;
 
 import static build.buildfarm.common.DigestUtil.HashFunction.SHA256;
+import static build.buildfarm.server.ByteStreamService.CHUNK_SIZE;
 import static com.google.common.truth.Truth.assertThat;
 import static org.mockito.AdditionalAnswers.answerVoid;
 import static java.util.concurrent.TimeUnit.SECONDS;
@@ -35,11 +36,15 @@ import build.buildfarm.common.Write;
 import build.buildfarm.common.io.FeedbackOutputStream;
 import build.buildfarm.instance.Instance;
 import build.buildfarm.instance.stub.ByteStreamUploader;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
 import com.google.common.hash.HashCode;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.SettableFuture;
 import com.google.bytestream.ByteStreamGrpc;
 import com.google.bytestream.ByteStreamGrpc.ByteStreamStub;
+import com.google.bytestream.ByteStreamProto.ReadRequest;
+import com.google.bytestream.ByteStreamProto.ReadResponse;
 import com.google.bytestream.ByteStreamProto.WriteRequest;
 import com.google.bytestream.ByteStreamProto.WriteResponse;
 import com.google.protobuf.ByteString;
@@ -48,8 +53,10 @@ import io.grpc.Server;
 import io.grpc.Status;
 import io.grpc.inprocess.InProcessChannelBuilder;
 import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.ServerCallStreamObserver;
 import io.grpc.stub.StreamObserver;
 import java.io.IOException;
+import java.util.List;
 import java.util.UUID;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
@@ -59,6 +66,7 @@ import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 import org.mockito.invocation.InvocationOnMock;
@@ -255,5 +263,93 @@ public class ByteStreamServiceTest {
     verify(write, atLeastOnce()).getCommittedSize();
     verify(write, atLeastOnce()).getOutput(any(Long.class), any(TimeUnit.class), any(Runnable.class));
     verify(write, times(2)).addListener(any(Runnable.class), any(Executor.class));
+  }
+
+  class CountingReadObserver implements StreamObserver<ReadResponse> {
+    private final List<Integer> sizes = Lists.newArrayList();
+    private final ByteString.Output sink = ByteString.newOutput();
+    private boolean completed = false;
+
+    @Override
+    public void onNext(ReadResponse response) {
+      ByteString data = response.getData();
+      try {
+        data.writeTo(sink);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      sizes.add(data.size());
+    }
+
+    @Override
+    public void onCompleted() {
+      completed = true;
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      t.printStackTrace();
+    }
+
+    List<Integer> getSizesList() {
+      return sizes;
+    }
+
+    boolean isCompleted() {
+      return completed;
+    }
+
+    ByteString getData() {
+      return sink.toByteString();
+    }
+  }
+
+  @Test
+  public void readSlicesLargeChunksFromInstance() throws Exception {
+    // pick a large chunk size
+    long size = CHUNK_SIZE * 10 + CHUNK_SIZE - 47;
+    ByteString content;
+    try (ByteString.Output out = ByteString.newOutput(ByteStreamService.CHUNK_SIZE * 10 + ByteStreamService.CHUNK_SIZE - 47)) {
+      for (long i = 0; i < size; i++) {
+        out.write((int) (i & 0xff));
+      }
+      content = out.toByteString();
+    }
+    Digest digest = DIGEST_UTIL.compute(content);
+    String resourceName = "blobs/" + DigestUtil.toString(digest);
+    ReadRequest request = ReadRequest.newBuilder()
+        .setResourceName(resourceName)
+        .build();
+
+    Instance instance = mock(Instance.class);
+    when(instances.getFromBlob(eq(resourceName))).thenReturn(instance);
+    doAnswer(answerVoid((blobDigest, offset, limit, chunkObserver, metadata) -> {})).when(instance).getBlob(
+        eq(digest),
+        eq(request.getReadOffset()),
+        eq((long) content.size()),
+        any(ServerCallStreamObserver.class),
+        eq(RequestMetadata.getDefaultInstance()));
+    Channel channel = InProcessChannelBuilder.forName(fakeServerName).directExecutor().build();
+    ByteStreamStub service = ByteStreamGrpc.newStub(channel);
+    CountingReadObserver readObserver = new CountingReadObserver();
+    service.read(request, readObserver);
+    ArgumentCaptor<ServerCallStreamObserver<ByteString>> observerCaptor = ArgumentCaptor.forClass(ServerCallStreamObserver.class);
+    verify(instance, times(1)).getBlob(
+        eq(digest),
+        eq(request.getReadOffset()),
+        eq((long) content.size()),
+        observerCaptor.capture(),
+        eq(RequestMetadata.getDefaultInstance()));
+
+    StreamObserver<ByteString> responseObserver = observerCaptor.getValue();
+    // supply entire content
+    responseObserver.onNext(content);
+    responseObserver.onCompleted();
+
+    assertThat(readObserver.isCompleted()).isTrue();
+    assertThat(readObserver.getData()).isEqualTo(content);
+    List<Integer> sizes = readObserver.getSizesList();
+    assertThat(sizes.size()).isEqualTo(11); // 10 + 1 incomplete chunk
+    assertThat(Iterables.filter(sizes, (responseSize) -> responseSize > CHUNK_SIZE)).isEmpty();
   }
 }

